src/xma/src/xmaplugin/xmaplugin.cpp

```diff
@@ -31,32 +31,60 @@ using namespace std;
 
 #define XMAPLUGIN_MOD "xmapluginlib"
 
-//Sarab: TODO .. Assign buffr object fields.. bank etc..
-//Add new API for allocating device only buffer object.. which will not have host mappped buffer.. This could be used for zero copy plugins..
-//NULL data pointer in buffer obj implies it is device only buffer..
-//Remove read/write before SyncBO.. As plugin should manage that using host mapped data pointer..
+
+int32_t create_bo(xclDeviceHandle dev_handle, XmaBufferObj& b_obj, uint32_t size, uint32_t ddr_bank, 
+    bool device_only_buffer, xclBufferHandle& b_obj_handle) {
+    /*
+    #define XRT_BO_FLAGS_MEMIDX_MASK        (0xFFFFFFUL)
+    #define XCL_BO_FLAGS_CACHEABLE          (1 << 24)
+    #define XCL_BO_FLAGS_SVM                (1 << 27)
+    #define XCL_BO_FLAGS_DEV_ONLY           (1 << 28)
+    #define XCL_BO_FLAGS_HOST_ONLY          (1 << 29)
+    #define XCL_BO_FLAGS_P2P                (1 << 30)
+    #define XCL_BO_FLAGS_EXECBUF            (1 << 31)
+    */
+    if (device_only_buffer) {
+        b_obj_handle = xclAllocBO(dev_handle, size, 0, XCL_BO_FLAGS_DEV_ONLY | ddr_bank);
+        b_obj.device_only_buffer = true;
+    } else {
+        b_obj_handle = xclAllocBO(dev_handle, size, 0, ddr_bank);
+    }
+
+    struct xclBOProperties bop;
+    if (xclGetBOProperties(dev_handle, b_obj_handle, &bop) != 0) {
+        xma_logmsg(XMA_ERROR_LOG, XMAPLUGIN_MOD, "xma_plg_buffer_alloc failed to get BO properties\n");
+        xclFreeBO(dev_handle, b_obj_handle);
+        return XMA_ERROR;
+    }
+    b_obj.paddr = bop.paddr;
+
+    if (!device_only_buffer) {
+        b_obj.data = (uint8_t*) xclMapBO(dev_handle, b_obj_handle, true);
+    }
+    return XMA_SUCCESS;
+}
 
 XmaBufferObj
 xma_plg_buffer_alloc(XmaSession s_handle, size_t size, bool device_only_buffer, int32_t* return_code)
 {
     XmaBufferObj b_obj;
     XmaBufferObj b_obj_error;
-    b_obj_error.data = NULL;
+    b_obj_error.data = nullptr;
```

Comment:
Thank you ... and I am sure Soren thanks you too :^)

Reply:
:)